buffer/src/main/java/io/netty/buffer/CompositeByteBuf.java

```diff
@@ -1519,54 +1585,53 @@ public ByteBuffer nioBuffer(int index, int length) {
             return new ByteBuffer[] { EMPTY_NIO_BUFFER };
         }
 
-        List<ByteBuffer> buffers = new ArrayList<ByteBuffer>(components.size());
-        int i = toComponentIndex(index);
-        while (length > 0) {
-            Component c = components.get(i);
-            ByteBuf s = c.buf;
-            int adjustment = c.offset;
-            int localLength = Math.min(length, s.capacity() - (index - adjustment));
-            switch (s.nioBufferCount()) {
+        RecyclableArrayList buffers = RecyclableArrayList.newInstance(componentCount);
+        try {
+            int i = toComponentIndex0(index);
+            while (length > 0) {
+                Component c = components[i];
+                ByteBuf s = c.buf;
+                int localLength = Math.min(length, c.endOffset - index);
+                switch (s.nioBufferCount()) {
                 case 0:
                     throw new UnsupportedOperationException();
                 case 1:
-                    buffers.add(s.nioBuffer(index - adjustment, localLength));
+                    buffers.add(s.nioBuffer(c.idx(index), localLength));
                     break;
                 default:
-                    Collections.addAll(buffers, s.nioBuffers(index - adjustment, localLength));
+                    Collections.addAll(buffers, s.nioBuffers(c.idx(index), localLength));
+                }
+
+                index += localLength;
+                length -= localLength;
+                i ++;
             }
 
-            index += localLength;
-            length -= localLength;
-            i ++;
+            return buffers.toArray(new ByteBuffer[0]);
```

Comment:
should we cache an empty array?

Reply:
There was some discussion about this general question [here](https://github.com/netty/netty/pull/8075#issuecomment-401143387), I think the conclusion (and current precedent elsewhere in the netty code) is to _not_ cache such empty arrays.

Reply:
@mosesn I think we dont need to and we also dont do in many other places.